rclcpp/include/rclcpp/node_handle.hpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__NODE_HANDLE_HPP_
+#define RCLCPP__NODE_HANDLE_HPP_
+
+#include <memory>
+#include <type_traits>
+
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_clock_interface.hpp"
+#include "rclcpp/node_interfaces/node_graph_interface.hpp"
+#include "rclcpp/node_interfaces/node_logging_interface.hpp"
+#include "rclcpp/node_interfaces/node_parameters_interface.hpp"
+#include "rclcpp/node_interfaces/node_services_interface.hpp"
+#include "rclcpp/node_interfaces/node_time_source_interface.hpp"
+#include "rclcpp/node_interfaces/node_timers_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/node_interfaces/node_waitables_interface.hpp"
+
+namespace rclcpp
+{
+
+using BaseInterface = node_interfaces::NodeBaseInterface::SharedPtr;
+using ClockInterface = node_interfaces::NodeClockInterface::SharedPtr;
+using GraphInterface = node_interfaces::NodeGraphInterface::SharedPtr;
+using LoggingInterface = node_interfaces::NodeLoggingInterface::SharedPtr;
+using TimersInterface = node_interfaces::NodeTimersInterface::SharedPtr;
+using TopicsInterface = node_interfaces::NodeTopicsInterface::SharedPtr;
+using ServicesInterface = node_interfaces::NodeServicesInterface::SharedPtr;
+using WaitablesInterface = node_interfaces::NodeWaitablesInterface::SharedPtr;
+using ParametersInterface = node_interfaces::NodeParametersInterface::SharedPtr;
+using TimeSourceInterface = node_interfaces::NodeTimeSourceInterface::SharedPtr;
+struct AllInterfaces {};
+
+/// A helper class for aggregating node interfaces
+template<typename ... InterfaceTs>
+class NodeHandle : public std::enable_shared_from_this<NodeHandle<InterfaceTs...>>
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(NodeHandle)
+
+  /// Create a new node handle with no bound node interfaces.
+  RCLCPP_PUBLIC
+  NodeHandle()
+  : base_(nullptr),
+    clock_(nullptr),
+    graph_(nullptr),
+    logging_(nullptr),
+    timers_(nullptr),
+    topics_(nullptr),
+    services_(nullptr),
+    waitables_(nullptr),
+    parameters_(nullptr),
+    time_source_(nullptr)
```

Comment:
I think we should talk about this design. I have a few issues with it, which maybe we can improve on:

- it copies and stores the interface pointers itself, when instead we could just keep a pointer to the node class itself (which already has these pointers)
- this list is finite and has to be adjusted each time a new interface is added. I'm not sure we want to have that, and I think if we address the previous point we can avoid having to have an exhaustive list of interfaces in this class

Reply:
I was thinking of storing a node pointer to the node, but I decided against it for two reasons (though the latter is probably more significant than the former):
- It'll introduce an additional layer of function indirection
- We won't be able to aggregate node interfaces together (making them from parts)

I think the latter, aggregating reason is more important because it was one of the requirements I was trying to fulfill [from the discussion in the motivating issue](https://github.com/ros2/rclcpp/issues/831#issuecomment-530157701). But it's true that it means we'll need to adjust the NodeHandle class each time a new interface is added, though in my opinion it's a worthwhile trade-off since that shouldn't happen as often, I think?

Reply:
I agree with @methylDragon, we should allow this class to be constructed from generic objects, i.e not necessarily rclcpp::node or rclcpp_lifecycle::LifecycleNode

Reply:
I like this idea, it'll allow for users to create their own interfaces and bind them to the node handle as well.

Only thing is that this will also allow binding and inheritance from any possible class (as long as it has a constructor that takes nodeT). I think this is a fair enough tradeoff, since it also resolves @alsora 's

> Can we implement it in a way such that users can customize it? I.e. If I create a new node interface, I would like to add it there

And additionally gives me a good reason for introducing more terse type names for use with the template args.

---

I'll also move the implementation to the node_interfaces directory